File: R/features.R
# Feature engineering helpers.
#
# Two families of features are supported:
#   1. `tfidf_recipe()`  — a textrecipes recipe that tokenizes and computes
#      TF-IDF inside the modeling workflow, so the vocabulary is learned only
#      from training folds (no leakage during cross-validation).
#   2. `stylometric_features()` — simple hand-crafted features (sentence
#      length, punctuation rates, type-token ratio, etc.) that the team can
#      use as descriptive variables or as additional inputs to a classifier.

suppressPackageStartupMessages({
  library(recipes)
  library(textrecipes)
  library(dplyr)
  library(stringr)
  library(tibble)
})

#' Build a TF-IDF recipe with sensible defaults for essay-length text.
#'
#' @param formula A formula such as `generated ~ text`.
#' @param data Training data used to declare the recipe roles.
#' @param max_tokens Vocabulary cap after token filtering.
#' @param min_times Minimum number of essays a token must appear in.
#' @param ngram Maximum n-gram length (1 = unigrams only, 2 = uni+bigrams).
#' @return A `recipes::recipe` object.
tfidf_recipe <- function(formula, data,
                         max_tokens = 20000L,
                         min_times = 2L,
                         ngram = 2L) {
  rec <- recipes::recipe(formula, data = data) |>
    textrecipes::step_tokenize(text) |>
    textrecipes::step_stopwords(text)

  if (ngram > 1) {
    rec <- rec |> textrecipes::step_ngram(text, num_tokens = ngram, min_num_tokens = 1L)
  }

  rec |>
    textrecipes::step_tokenfilter(text, max_tokens = max_tokens, min_times = min_times) |>
    textrecipes::step_tfidf(text)
}

#' Compute readability scores for each essay.
#'
#' Wraps `quanteda.textstats::textstat_readability()` so that the team can
#' use industry-standard readability indices as additional descriptive or
#' predictive features.
#'
#' @param texts Character vector of essays.
#' @param measures Which indices to compute. Defaults to Flesch reading ease,
#'   Flesch-Kincaid grade level, and the Automated Readability Index.
#' @return A tibble with one row per essay and one column per measure.
readability_features <- function(texts,
                                 measures = c("Flesch", "Flesch.Kincaid", "ARI")) {
  if (!requireNamespace("quanteda.textstats", quietly = TRUE)) {
    stop("Package 'quanteda.textstats' is required. Run setup.R to install it.")
  }
  out <- quanteda.textstats::textstat_readability(
    as.character(texts),
    measure = measures
  )
  tibble::as_tibble(out[, measures, drop = FALSE])
}

#' Compute a small table of hand-crafted stylometric features.
#'
#' @param texts Character vector of essays.
#' @param include_readability If TRUE, also compute Flesch / Flesch-Kincaid / ARI.
#' @return A tibble with one row per essay.
stylometric_features <- function(texts, include_readability = FALSE) {
  texts <- as.character(texts)

  word_lists <- stringr::str_extract_all(texts, "\\b\\w+\\b")
  sentence_lists <- stringr::str_split(texts, "[.!?]+")
  sentence_lists <- lapply(sentence_lists, function(s) s[nchar(trimws(s)) > 0])

  n_chars <- pmax(nchar(texts), 1L)
  n_words <- pmax(lengths(word_lists), 1L)
  n_sentences <- pmax(lengths(sentence_lists), 1L)

  avg_word_len <- vapply(word_lists, function(w) {
    if (length(w) == 0) 0 else mean(nchar(w))
  }, numeric(1))

  unique_word_count <- vapply(word_lists, function(w) length(unique(tolower(w))), integer(1))

  punct_count <- stringr::str_count(texts, "[[:punct:]]")
  comma_count <- stringr::str_count(texts, ",")
  upper_count <- stringr::str_count(texts, "[A-Z]")

  out <- tibble::tibble(
    n_chars = as.numeric(nchar(texts)),
    n_words = as.numeric(lengths(word_lists)),
    n_sentences = as.numeric(lengths(sentence_lists)),
    avg_word_len = avg_word_len,
    avg_sentence_len_words = n_words / n_sentences,
    type_token_ratio = unique_word_count / n_words,
    punct_rate = punct_count / n_chars,
    comma_rate = comma_count / n_chars,
    uppercase_rate = upper_count / n_chars
  )

  if (isTRUE(include_readability)) {
    out <- dplyr::bind_cols(out, readability_features(texts))
  }
  out
}

File: src/features.py
"""Feature engineering helpers.

Two families of features are supported:

1. ``build_tfidf_vectorizer`` — a ready-to-use TF-IDF vectorizer with
   sensible defaults for essay-length text.
2. ``stylometric_features`` — simple hand-crafted features (sentence
   length, punctuation rates, type-token ratio, etc.) that the team can use
   as descriptive variables or as additional inputs to a classifier.
"""

from __future__ import annotations

import re
import string
from typing import Iterable

import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer

_SENTENCE_SPLIT = re.compile(r"[.!?]+")
_WORD_SPLIT = re.compile(r"\b\w+\b")


def build_tfidf_vectorizer(
    *,
    max_features: int = 20_000,
    ngram_range: tuple[int, int] = (1, 2),
    min_df: int = 2,
    max_df: float = 0.95,
    sublinear_tf: bool = True,
) -> TfidfVectorizer:
    """Return a TfidfVectorizer with defaults tuned for essay text."""
    return TfidfVectorizer(
        lowercase=True,
        strip_accents="unicode",
        max_features=max_features,
        ngram_range=ngram_range,
        min_df=min_df,
        max_df=max_df,
        sublinear_tf=sublinear_tf,
    )


def _stylometric_row(text: str) -> dict[str, float]:
    words = _WORD_SPLIT.findall(text)
    sentences = [s for s in _SENTENCE_SPLIT.split(text) if s.strip()]
    n_words = len(words) or 1
    n_sentences = len(sentences) or 1
    n_chars = len(text) or 1

    punct_count = sum(1 for c in text if c in string.punctuation)
    unique_words = {w.lower() for w in words}

    return {
        "n_chars": float(len(text)),
        "n_words": float(len(words)),
        "n_sentences": float(len(sentences)),
        "avg_word_len": float(np.mean([len(w) for w in words])) if words else 0.0,
        "avg_sentence_len_words": n_words / n_sentences,
        "type_token_ratio": len(unique_words) / n_words,
        "punct_rate": punct_count / n_chars,
        "comma_rate": text.count(",") / n_chars,
        "uppercase_rate": sum(1 for c in text if c.isupper()) / n_chars,
    }


def stylometric_features(texts: Iterable[str]) -> pd.DataFrame:
    """Compute a small table of hand-crafted stylometric features."""
    return pd.DataFrame([_stylometric_row(t) for t in texts])
